soroban-env-host/src/crypto/bls12_381.rs

```diff
@@ -0,0 +1,665 @@
+            for p in vp.iter() {
+                let pp =
+                    self.g1_affine_deserialize_from_bytesobj(BytesObject::try_from_val(self, p)?)?;
+                points.push(pp);
+            }
+            Ok(())
+        });
+        Ok(points)
+    }
+
+    pub(crate) fn scalar_vec_from_vecobj(&self, vs: VecObject) -> Result<Vec<Fr>, HostError> {
+        let len: u32 = self.vec_len(vs)?.into();
+        let mut scalars: Vec<Fr> = vec![];
+        self.charge_budget(
+            ContractCostType::MemAlloc,
+            Some(len as u64 * FR_SERIALIZED_SIZE as u64),
+        )?;
+        scalars.reserve(len as usize);
+        let _ = self.visit_obj(vs, |vs: &HostVec| {
+            for s in vs.iter() {
+                let ss = self.fr_from_u256val(U256Val::try_from_val(self, s)?)?;
+                scalars.push(ss);
+            }
+            Ok(())
+        });
+        Ok(scalars)
+    }
+
+    pub(crate) fn g1_add_internal(
+        &self,
+        p0: G1Affine,
+        p1: G1Affine,
+    ) -> Result<G1Projective, HostError> {
+        self.charge_budget(ContractCostType::Bls12381G1Add, None)?;
+        Ok(p0.add(p1))
+    }
+
+    pub(crate) fn g1_mul_internal(
+        &self,
+        p0: G1Affine,
+        scalar: Fr,
+    ) -> Result<G1Projective, HostError> {
+        self.charge_budget(ContractCostType::Bls12381G1Mul, None)?;
+        Ok(p0.mul(scalar))
+    }
+
+    pub(crate) fn g1_msm_internal(
+        &self,
+        points: &[G1Affine],
+        scalars: &[Fr],
+    ) -> Result<G1Projective, HostError> {
+        // this check should've been done outside, here is just extra caution
+        if points.len() != scalars.len() || points.len() == 0 {
+            return Err(
+                Error::from_type_and_code(ScErrorType::Crypto, ScErrorCode::InvalidInput).into(),
+            );
+        }
+        // The actual logic happens inside msm_bigint_wnaf (ark_ec/variable_base/mod.rs)
+        // under branch negation is cheap.
+        // the unchecked version just skips the length equal check
+        self.charge_budget(ContractCostType::Bls12381G1Msm, Some(points.len() as u64))?;
+        Ok(G1Projective::msm_unchecked(points, scalars))
+    }
+
+    pub(crate) fn map_fp_to_g1_internal(&self, fp: Fq) -> Result<G1Affine, HostError> {
+        self.charge_budget(ContractCostType::Bls12381MapFpToG1, None)?;
+        let mapper = WBMap::<g1::Config>::new().map_err(|e| {
+            self.err(
+                ScErrorType::Crypto,
+                ScErrorCode::InternalError,
+                format!("hash-to-curve error {e}").as_str(),
+                &[],
+            )
+        })?;
+        mapper.map_to_curve(fp).map_err(|e| {
+            self.err(
+                ScErrorType::Crypto,
+                ScErrorCode::InternalError,
+                format!("hash-to-curve error {e}").as_str(),
+                &[],
+            )
+        })
+    }
+
+    pub(crate) fn hash_to_g1_internal(
```

Comment:
+1 on removing code duplication.

Reply:
I'm all for reducing code size. And most of the helper routines has already been templated/macroed whenever applicable.

However, generic code has maintenance cost, and it less readable and documentable. A lot of these higher level routines mentioned are only a few statements long, which already call into the lower level generic functions. 
Converting these functions to further templates or macros will maybe introduce 4-5 parameters each, and the template/macro is only maybe 5-10 lines long. And it can only be reused twice (for g1 and g2). So I don't feel such slight code footprint saving is worth it for the additional development and maintenance cost. 

The arkworks code base is already highly generic (and not straightforward to navigate). For example, `Fq` and `Fq2` are two completely different generic types, one is the field element of a generic size `Fp<N>`, the other one is a quadratic_extension element. So I don't feel like generalizing those away is a good idea (also I wrote a long comment on their (de)serialization requirements), even though their end-product code looks like similar. 
 
Happy to reconsider it once the entire code is settled and stabilized. Maybe as a follow up. 
But for now I think adding more tests is at a much higher priority.

Reply:
I have done more refactoring to turn some of the more involved functions into templates. These include the deserialization, map-to-curve, hash-to-curve function. Please take another look and let me know if there are any other non-obvious ones.